source/fuzz/transformation_access_chain.cpp

```diff
@@ -196,21 +280,149 @@ protobufs::Transformation TransformationAccessChain::ToMessage() const {
 }
 
 std::pair<bool, uint32_t> TransformationAccessChain::GetIndexValue(
-    opt::IRContext* ir_context, uint32_t index_id) const {
+    opt::IRContext* ir_context, uint32_t index_id,
+    uint32_t object_type_id) const {
+  if (!ValidIndexToComposite(ir_context, index_id, object_type_id)) {
+    return {false, 0};
+  }
   auto index_instruction = ir_context->get_def_use_mgr()->GetDef(index_id);
-  if (!index_instruction || !spvOpcodeIsConstant(index_instruction->opcode())) {
-    // TODO(https://github.com/KhronosGroup/SPIRV-Tools/issues/3179) We could
-    //  allow non-constant indices when looking up non-structs, using clamping
-    //  to ensure they are in-bounds.
+
+  uint32_t bound = fuzzerutil::GetBoundForCompositeIndex(
+      *ir_context->get_def_use_mgr()->GetDef(object_type_id), ir_context);
+
+  // The index must be a constant
+  if (!spvOpcodeIsConstant(index_instruction->opcode())) {
+    return {false, 0};
+  }
+
+  // The index must be in bounds.
+  uint32_t value = index_instruction->GetSingleWordInOperand(0);
+
+  if (value >= bound) {
+    return {false, 0};
+  }
+
+  return {true, value};
+}
+
+std::pair<bool, uint32_t> TransformationAccessChain::CreateAndGetClampedIndexId(
+    opt::IRContext* ir_context, uint32_t index_id, uint32_t object_type_id,
+    bool add_clamping_instructions,
+    std::pair<uint32_t, uint32_t> fresh_ids) const {
+  if (!ValidIndexToComposite(ir_context, index_id, object_type_id)) {
     return {false, 0};
   }
+
+  // The composite must not be a struct
+  auto object_type_def = ir_context->get_def_use_mgr()->GetDef(object_type_id);
+
+  if (object_type_def->opcode() == SpvOpTypeStruct) {
+    return {false, 0};
+  }
+
+  // Valid fresh ids need to have been given
+  if (fresh_ids.first <= 0 || fresh_ids.second <= 0) {
+    return {false, 0};
+  }
```

Comment:
Again, in the "Apply" path this could be left out or asserted.

Reply:
Left out